dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/sam/SamPermissionsImpl.java

```diff
@@ -85,13 +131,14 @@ public SamPermissionsImpl(TokenDAO tokenDAO, DockstoreWebserviceConfiguration co
      * @param requester -- the requester, who must be an owner of <code>workflow</code> or an admin
      * @param workflow the workflow
      * @param permission -- the email and the permission for that email
-     * @return
+     * @return a list of permissions
+     * @throws CustomWebApplicationException
```

Comment:
thoughts on putting this into the method signature?
https://phauer.com/2015/checked-exceptions-are-evil/#documentation-of-unchecked-exceptions

Reply:
> thoughts on putting this into the method signature? https://phauer.com/2015/checked-exceptions-are-evil/#documentation-of-unchecked-exceptions

The Joshua Block "Effective Java" book says "Use the Javadoc @throws tag to document each exception that a method can throw, but do not use the throws keyword on unchecked exceptions"; your link argues the opposite. 🤔 

In general, it looks like we rarely declare `throws CustomWebapplicationException` in the codebase, but we sometimes do.

Reply:
What's the reasoning in Effective Java?

Reply:
It's a little vague... Unfortunately it's not open source, so I can't link to it directly, but here's some of the text:


> Use the Javadoc @throws tag to document each exception that a method can throw, but do not use the throws keyword on unchecked exceptions. It is important that programmers using your API are aware of which exceptions are checked and which are unchecked because the programmers’ responsibilities differ in these two cases. The documentation generated by the Javadoc @throws tag without a corresponding throws clause in the method declaration provides a strong visual cue to the programmer that an exception is unchecked.
> 
> It should be noted that documenting all of the unchecked exceptions that each method can throw is an ideal, not always achievable in the real world. When a class undergoes revision, it is not a violation of source or binary compatibility if an exported method is modified to throw additional unchecked exceptions. Suppose a class invokes a method from another, independently written class. The authors of the former class may carefully document all of the unchecked exceptions that each method throws, but if the latter class is revised to throw additional unchecked exceptions, it is quite likely that the former class (which has not undergone revision) will propagate the new unchecked exceptions even though it does not document them.

Reading it makes me think there should perhaps be a checked exception. Pondering...

Reply:
Well, I happen to agree with the "Checked Exceptions are Evil"  bit
I just have fewer feelings about whether to put it in the method signature (vs docs).

I do wonder though, the "The documentation generated by the Javadoc @throws tag without a corresponding throws clause in the method declaration provides a strong visual cue to the programmer that an exception is unchecked." bit may be out of date since the other article says "simply because you get better tooling support depending on your IDE (e.g. code completion when writing the exception in the catch block, highlighting of the throwing method if you select the exception in the catch block or vice versa). Moreover, you can still use JavaDoc additionally for further documentation about the exception."

Reply:
I started to declare it, but it had this big ripple effect, so I backed away. I updated the doc for the `email` field as way -- I don't think it's worth changing at this point, even though it is incorrect.

Reply:
By "declare it", do you mean add `throws CustomWebApplicationException` to the method signature, and if so, out of curiosity, what was the ripple effect?  Or did you attempt to use a checked exception (in which case, yeah, that would take a lot of adjusting...)?

Regarding "checked versus unchecked", IMHO, they each offer a certain behavior that's neither fundamentally good or bad, but useful in different situations.  I'm glad that `CustomWebApplicationException` is unchecked, because otherwise, so many `throws`, and we can use methods from which it emerges in the streams.  On the other hand, imagine some sort of mission critical API with errors that need to be apparent and immediately handled, checked exceptions might be just the ticket.

Reply:
> By "declare it", do you mean add `throws CustomWebApplicationException` to the method signature, and if so, out of curiosity, what was the ripple effect? Or did you attempt to use a checked exception (in which case, yeah, that would take a lot of adjusting...)?

Yeah, I was more thinking the former rather than the latter, so thinking less ripple. But meh

Reply:
> By "declare it", do you mean add `throws CustomWebApplicationException` to the method signature, and if so, out of curiosity, what was the ripple effect? Or did you attempt to use a checked exception (in which case, yeah, that would take a lot of adjusting...)?

The ripple effect is that this method implements an interface method, so then I should add the `throws` to the interface method declaration too. And then I started looking at other methods in the interface, and several of them can throw `CustomWebApplicationException` as well, so then to be consistent I should start adding the `throws` to all those methods, then delve into the methods' implementations (I know I wouldn't necessarily have to add the `throws` to all implementations, but still).

Reply:
ripplocalypse